### ISSUE TITLE
Docker container networking

### ISSUE THREAD

**Issue by dbrody2004** [netdata/netdata#5383]
We have a bunch of docker containers with names.  Is it possible to show the name instead of the container id?  Secondly, each of these containers have an veth interface and it would be nice to match those up to the container name - is that possible?

**Comment by cakrit** [comment]
> Is it possible to show the name instead of the container id?
Sure, start reading from https://github.com/netdata/netdata/issues/5314#issuecomment-460357316 onwards. 

> each of these containers have an veth interface and it would be nice to match those up to the container name - is that possible

I'll probably need some help from someone else on this one. So these interfaces do not appear as `net <interface identifier>` under each container id on the right menu? What exactly do you see in netdata?

**Comment by dbrody2004** [comment]
They show up as vethxxxxx under the interfaces tab and not under the docker container -

<img width="176" alt="screen shot 2019-02-13 at 4 32 55 pm" src="https://user-images.githubusercontent.com/4682241/52753833-1cc48680-2fad-11e9-9a4b-e8a60a5370ff.png">

<img width="180" alt="screen shot 2019-02-13 at 4 33 07 pm" src="https://user-images.githubusercontent.com/4682241/52753835-20580d80-2fad-11e9-9238-81ad936bc758.png">

**Comment by cakrit** [comment]
Right now, all you can do is what is explained [here](https://forums.docker.com/t/relationship-between-interface-vethxxxxx-and-container/12872/20) and with a script example [here](https://superuser.com/questions/1183454/finding-out-the-veth-interface-of-a-docker-container).

@vlvkobal, @ktsaou, @paulfantom it would be great if we could add that logic in netdata and be able to show these `veth` interfaces where they belong (under the container), but it's not clear to me how we'd do it.

**Comment by cakrit** [comment]
@ilyam8 found #2539, so this could be a bug. netdata should already be renaming those interfaces to move them to  the correct place.
 
I haven't read through the implementation yet, but I suspect that the issue of the interfaces could be resolved as soon as you resolve the issue with the container names not appearing (i.e. perhaps the match expects netdata to have the container name instead of the id). Did you check #5314  to get thae names working?

**Comment by cakrit** [comment]
I was wrong,  I was able to replicate it. Hold on while I debug...

**Comment by cakrit** [comment]
Ok, so to replicate the issue I followed @ilyam8 's advice and did the following:
```
sudo docker run \
  --volume=/:/rootfs:ro \
  --volume=/var/run:/var/run:ro \
  --volume=/sys:/sys:ro \
  --volume=/var/lib/docker/:/var/lib/docker:ro \
  --volume=/dev/disk/:/dev/disk:ro \
  --publish=8080:8080 \
  --detach=true \
  --name=cadvisor \
  google/cadvisor:latest
```
I looked at what netdata did. It called

`/usr/libexec/netdata/plugins.d/cgroup-network --cgroup '/sys/fs/cgroup/cpu,cpuacct/docker/1854044f202f231fa3121b40a0a6a06e6ab3c4537d5a0f89535a518936d069e4'`

which in turn called

`/usr/libexec/netdata/plugins.d/cgroup-network-helper.sh --cgroup '/sys/fs/cgroup/cpu,cpuacct/docker/1854044f202f231fa3121b40a0a6a06e6ab3c4537d5a0f89535a518936d069e4'`

The helper is the one that has a problem. It has two methods of operation, neither of which work in this case:
- Taking as an argument a pid, it looks for "iff" in `/proc/[PID]/fdinfo/*`. Those files are empty in these case.
- Taking as an argument the container path, it tries to use `virsh` (i.e. libvirt), with a domain that is supposedly contained in that path. However, `virsh -r list --all` is completely empty. I did some research on docker containers and `libvirt`, but I got in over my head and will need assistance.

An alternative method that seems to work is to do the following, when called with a container:

```
  container_id=1854044f202f231fa3121b40a0a6a06e6ab3c4537d5a0f89535a518936d069e4;
  pid=$(docker inspect -f '{{.State.Pid}}' ${container_id})
  mkdir -p /var/run/netns/
  ln -sfT /proc/$pid/ns/net /var/run/netns/$container_id
```
With these in place,  `ip link show` will give us the veth and the container id. 

I'm requesting @ktsaou to help here, additional insights are of course more than welcome.

**Comment by ktsaou** [comment]
I traced this issue. It seems like a kernel bug.

Let me explain it.

`cgroup-network` reads `/proc/net/dev`, then switches namespaces (to get inside the container) and reads `/proc/net/dev` again.

For each interface read (both at the host and container namespaces) it reads `/sys/class/net/{INTERFACE}/ifindex` and `/sys/class/net/{INTERFACE}/iflink`. These files contain a single number. Matching numbers (host ifindex = container iflink and vice versa) means the host interface is linked to the container interface.

To see the bug:

```sh
# find the pid of candvisor (as seen by the host)
[host] # pidof cadvisor
18761

# enter the namespace of that pid
[host] # nsenter -t 18761 -a /bin/sh
/ # 

# now you have a shell inside the container
# load the profile to update paths
/ # . /etc/profile
7403ca5f951d:/# 

# check the contents of /proc/net/dev
7403ca5f951d:/# cat /proc/net/dev
Inter-|   Receive                                                |  Transmit
 face |bytes    packets errs drop fifo frame compressed multicast|bytes    packets errs drop fifo colls carrier compressed
    lo:     500      10    0    0    0     0          0         0      500      10    0    0    0     0       0          0
  eth0:  194035     791    0    0    0     0          0         0        0       0    0    0    0     0       0          0

# they are good, eth0 is shown

# but this is the problem:

7403ca5f951d:/# ls -l /sys/class/net/
total 0
lrwxrwxrwx    1 root     root             0 Feb 18 09:04 br0 -> ../../devices/virtual/net/br0
lrwxrwxrwx    1 root     root             0 Feb 18 09:04 docker0 -> ../../devices/virtual/net/docker0
lrwxrwxrwx    1 root     root             0 Feb 18 09:04 enp3s0 -> ../../devices/pci0000:00/0000:00:1c.2/0000:03:00.0/net/enp3s0
lrwxrwxrwx    1 root     root             0 Feb 18 09:04 lo -> ../../devices/virtual/net/lo
lrwxrwxrwx    1 root     root             0 Feb 18 09:04 veth22ddf22 -> ../../devices/virtual/net/veth22ddf22
lrwxrwxrwx    1 root     root             0 Feb 18 09:04 virbr0 -> ../../devices/virtual/net/virbr0
lrwxrwxrwx    1 root     root             0 Feb 18 09:04 virbr0-nic -> ../../devices/virtual/net/virbr0-nic

```

So, `/sys/class/net` inside the container shows the host's interfaces, not the container ones. So, there is not `eth0` so `ifindex` and `iflink` cannot be read.

This is why `cgroup-network` logs this:

```sh
# grep cgroup-network /var/log/netdata/error.log
2019-02-18 09:49:40: /usr/libexec/netdata/plugins.d/cgroup-network INFO  : MAIN : running: exec /usr/libexec/netdata/plugins.d/cgroup-network-helper.sh --cgroup '/sys/fs/cgroup/cpu,cpuacct/docker/7403ca5f951d59cbd8bb18d111b282e289c2d6e38fd2d233302d2454c996d3dd'
2019-02-18 09:49:40: cgroup-network-helper.sh: INFO: searching for network interfaces of cgroup '/sys/fs/cgroup/cpu,cpuacct/docker/7403ca5f951d59cbd8bb18d111b282e289c2d6e38fd2d233302d2454c996d3dd'
2019-02-18 09:49:40: /usr/libexec/netdata/plugins.d/cgroup-network ERROR : MAIN : child pid 18928 exited with code 1.
2019-02-18 09:49:40: /usr/libexec/netdata/plugins.d/cgroup-network ERROR : MAIN : Cannot read '/sys/class/net/eth0/ifindex'. (errno 2, No such file or directory)
2019-02-18 09:49:40: /usr/libexec/netdata/plugins.d/cgroup-network ERROR : MAIN : Cannot read '/sys/class/net/eth0/iflink'. (errno 2, No such file or directory)
2019-02-18 09:49:40: /usr/libexec/netdata/plugins.d/cgroup-network ERROR : MAIN : there are not double-linked cgroup interfaces available.
```

**Comment by ktsaou** [comment]
On the same host, I started an lxc container. `/sys/class/net` seems to be right in this case.

**Comment by ktsaou** [comment]
hm... the problem is the docker command line:

```sh
sudo docker run \
  --volume=/:/rootfs:ro \
  --volume=/var/run:/var/run:ro \
  --volume=/sys:/sys:ro \
  --volume=/var/lib/docker/:/var/lib/docker:ro \
  --volume=/dev/disk/:/dev/disk:ro \
  --publish=8080:8080 \
  --detach=true \
  --name=cadvisor \
  google/cadvisor:latest
```

The parameter `--volume=/sys:/sys:ro` maps the host's `/sys` to the container's `/sys`. So, `cgroup-network` cannot find the interfaces of the container!

**Comment by ktsaou** [comment]
Another way to find the network interface is this:

```sh
# get the namespace id
# lsns -t net -p $(pidof cadvisor) -o NETNSID -nr
0

# then use that number like this:
# ip link show | grep -B 1 "link-netnsid 0"
9: veth22ddf22@if8: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master docker0 state UP mode DEFAULT group default 
    link/ether 06:d3:7c:90:eb:d8 brd ff:ff:ff:ff:ff:ff link-netnsid 0
```

**Comment by ktsaou** [comment]
I implemented to above method with PR #5413 
So, even when `ifindex` and `iflink` are missing, netdata can still detect the network interfaces that belong to a container. However, in this case the name of the network interface inside the container cannot be detected.
